File: 340.Longest_Substring_with_At_Most_K_Distinct_Characters/lengthOfLongestSubstringKDistinct.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
    var map = new Map();
    var low = 0;
    var len = 0;
    for(var i = 0; i < s.length; i++){
        map.set(s[i], i);
        if (map.size > k) {
            var smallest = Number.MAX_SAFE_INTEGER;
            //get the smallest value
            for (var value of map.values()) {
              smallest = Math.min(smallest, value);
            }
            low = smallest;
            map.delete(s[low]);
            low++;
        }
        len = Math.max(len, i - low + 1);
    }
    return len;
};

console.log(lengthOfLongestSubstringKDistinct("eceba",2));

File: 340.Longest_Substring_with_At_Most_K_Distinct_Characters/lengthOfLongestSubstringKDistinct.py
'''
Given a string, find the length of the longest substring T that contains at most k distinct characters.

For example, Given s = “eceba” and k = 2,

T is "ece" which its length is 3.
'''


class Solution(object):
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        res, low = 0, 0
        dic = {}
        for i in xrange(len(s)):
            dic[s[i]] = i
            if len(dic) > k:
                low = min(dic.values())
                del dic[s[low]]
                low += 1
            res = max(res, i - low + 1)
        return res


'''
Time complexity of this is O(nk), because it takes O(k) to find the low index
'''
